File: javascript/bird-watcher/bird-watcher.js
// @ts-check
//
// The line above enables type checking for this file. Various IDEs interpret
// the @ts-check directive. It will give you helpful autocompletion when
// implementing this exercise.

/**
 * Calculates the total bird count.
 *
 * @param {number[]} birdsPerDay
 * @returns {number} total bird count
 */
export function totalBirdCount(birdsPerDay) {
  //throw new Error('Please implement the totalBirdCount function');
  let count = 0;
  for (let index = 0; index < birdsPerDay.length; index++) {
    const element = birdsPerDay[index];
    count += element;
  }
  return count;
}

/**
 * Calculates the total number of birds seen in a specific week.
 *
 * @param {number[]} birdsPerDay
 * @param {number} week
 * @returns {number} birds counted in the given week
 */
export function birdsInWeek(birdsPerDay, week) {
  //throw new Error('Please implement the birdsInWeek function');
  const weekLength = 7;
  let weekStart = (week-1) * 7;
  let weekEnd = weekStart + 7;
  let count = 0;

  for (let index = weekStart; index < weekEnd; index++) {
    //const element = birdsPerDay[index];
    count += birdsPerDay[index];
  }
  return count;
}

/**
 * Fixes the counting mistake by increasing the bird count
 * by one for every second day.
 *
 * @param {number[]} birdsPerDay
 * @returns {number[]} corrected bird count data
 */
export function fixBirdCountLog(birdsPerDay) {
  //throw new Error('Please implement the fixBirdCountLog function');

  //add 1 to count for every second day
  for (let index = 0; index < birdsPerDay.length; index+=2) {
    birdsPerDay[index] += 1;
  }
  return birdsPerDay;

}

File: go/bird-watcher/bird_watcher.go
package birdwatcher

// TotalBirdCount return the total bird count by summing
// the individual day's counts.
func TotalBirdCount(birdsPerDay []int) int {
	count := 0
	for i := 0; i < len(birdsPerDay); i++ {
		count += birdsPerDay[i]
	}
	return count
}

// BirdsInWeek returns the total bird count by summing
// only the items belonging to the given week.
func BirdsInWeek(birdsPerDay []int, week int) int {
	const weekLength = 7
	weekStart := (week - 1) * 7
	weekEnd := weekStart + 7
	count := 0

	for i := weekStart; i < weekEnd; i++ {
		count += birdsPerDay[i]
	}
	return count
}

// FixBirdCountLog returns the bird counts after correcting
// the bird counts for alternate days.
func FixBirdCountLog(birdsPerDay []int) []int {

	// add 1 bird to count for every second day
	for i := 0; i < len(birdsPerDay); i += 2 {
		birdsPerDay[i] += 1
	}
	return birdsPerDay
}
